tiledb/api/src/query/conditions.rs

```diff
@@ -0,0 +1,754 @@
+            )
+        })?;
+
+        Ok(cond)
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum Condition {
+    Cmp(CmpCondition),
+    SetMember(SetMemberCondition),
+    Null(NullnessCondition),
+}
+
+impl Condition {
+    fn build<'ctx>(
+        &self,
+        ctx: &'ctx Context,
+    ) -> TileDBResult<QueryCondition<'ctx>> {
+        match self {
+            Self::Cmp(cond) => cond.build(ctx),
+            Self::SetMember(cond) => cond.build(ctx),
+            Self::Null(cond) => cond.build(ctx),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct Field {
+    field: String,
+}
+
+impl Field {
+    pub fn lt<V: Into<CmpValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::Cmp(CmpCondition {
+            field: self.field,
+            op: CmpOp::Less,
+            value: value.into(),
+        }))
+    }
+
+    pub fn le<V: Into<CmpValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::Cmp(CmpCondition {
+            field: self.field,
+            op: CmpOp::LessEqual,
+            value: value.into(),
+        }))
+    }
+
+    pub fn eq<V: Into<CmpValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::Cmp(CmpCondition {
+            field: self.field,
+            op: CmpOp::Equal,
+            value: value.into(),
+        }))
+    }
+
+    pub fn ne<V: Into<CmpValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::Cmp(CmpCondition {
+            field: self.field,
+            op: CmpOp::NotEqual,
+            value: value.into(),
+        }))
+    }
+
+    pub fn ge<V: Into<CmpValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::Cmp(CmpCondition {
+            field: self.field,
+            op: CmpOp::GreaterEqual,
+            value: value.into(),
+        }))
+    }
+
+    pub fn gt<V: Into<CmpValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::Cmp(CmpCondition {
+            field: self.field,
+            op: CmpOp::Greater,
+            value: value.into(),
+        }))
+    }
+
+    pub fn is_in<V: Into<SetMemberValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::SetMember(SetMemberCondition {
+            field: self.field,
+            op: SetMemberOp::In,
+            value: value.into(),
+        }))
+    }
+
+    pub fn not_in<V: Into<SetMemberValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::SetMember(SetMemberCondition {
+            field: self.field,
+            op: SetMemberOp::NotIn,
+            value: value.into(),
+        }))
+    }
+
+    pub fn is_null(self) -> QC {
+        QC::Cond(Condition::Null(NullnessCondition {
+            field: self.field,
+            op: NullnessOp::IsNull,
+        }))
+    }
+
+    pub fn not_null(self) -> QC {
+        QC::Cond(Condition::Null(NullnessCondition {
+            field: self.field,
+            op: NullnessOp::NotNull,
+        }))
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum QC {
```

Comment:
I really don't like "QC" as a name here, if the examples want to re-export a name as `QC` then that seems fine but I definitely feel like the proper type name should be more verbose.

Is there a reason you didn't move the `Condition` variants into this, and just have one enum?

Reply:
That's fair. I'll write out the QueryCondition name for the implementation and then provide a re-exported `QC` alias.

Reply:
Ooops. I went with `QueryConditionExpr` because of the name clash with the actual `QueryCondition`.